compensations/src/main/java/org/jboss/narayana/compensations/impl/ParticipantImpl.java

```diff
@@ -92,9 +94,10 @@ public void close() throws WrongStateException, SystemException {
             ConfirmationHandler handler = instantiate(confirmationHandler);
             handler.confirm();
 
-            CompensationContext.close(currentTX);
             Thread.currentThread().setContextClassLoader(origClassLoader);
         }
+
+        CompensationContextLifetimeManager.getInstance().participantDestroyed(currentTX);
```

Comment:
I refer to the methods in ParticipantImpl. For example:
public void close() throws WrongStateException, SystemException

If the body of the method throws one of these exceptions then your participantDestroyed call is not invoked.